### PR TITLE
User id replace level

### PR DESCRIPTION
**Pull request by Bjvanminnen**
This creates a new level type UserIdReplace (open to feedback on the name) that is essentially just an External level where we'll replace instances of a magic token with the current user's id. If we don't have a current user, it just strips the token (though the particular places we plan to use this will be scripts where you must be signed in).

The current magic token is `<user_id/>`. Also open to feedback on whether we need something more "magic".

### CODE REVIEW

File: dashboard/app/dsl/user_id_replace_dsl.rb [hunk 1]
```diff
@@ -0,0 +1,2 @@
+class UserIdReplaceDSL < ExternalDSL
+end
```
Comment (joshlory) [inline dashboard/app/dsl/user_id_replace_dsl.rb:2]: It's reasonable to follow the existing pattern in `ExternalDSL`, but given that we're trying to move away from `DSLDefined` I wonder if it makes sense to use the `.level` format?  Same feedback as here: https://github.com/code-dot-org/code-dot-org/pull/13833#discussion_r106543404

Comment (Bjvanminnen) [reply dashboard/app/dsl/user_id_replace_dsl.rb:2]: The goal of this was to be just like External, with this added functionality. We probably could just add this functionality to External itself and be fine. Not using External would essentially meaning porting External to a non DSLDefine type, which might be goodness, but is a lot more work and out of scope.

Comment (Bjvanminnen) [reply dashboard/app/dsl/user_id_replace_dsl.rb:2]: #13833 was a different enough level type that I was okay with creating a new type - in this case we want exactly the behavior of External, with this one extra little thing (replacing user_id).

File: dashboard/app/models/user_id_replace.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+#  index_levels_on_game_id  (game_id)
+#  index_levels_on_name     (name)
+#
+
+# This is a simple extension of the External level type that looks for the string
+# <user_id/> in haml and replaces it with the user's id
+class UserIdReplace < External
+  def properties_with_replaced_markdown(user)
+    user_id = user.try(:id).to_s
+    properties.merge({'markdown' => properties['markdown'].gsub('<user_id/>', user_id)})
```
Comment (joshlory) [inline dashboard/app/models/user_id_replace.rb:31]: Nit: magic string `<user_id/>` should maybe be a const (re-used in unit tests below).

**Comment by joshlory** [comment]
This type is similar enough to `External` it makes me wonder if we should just add this functionality to the existing `External` level type (after searching for any existing instances of `<user_id/>`).

File: dashboard/app/dsl/user_id_replace_dsl.rb (hunk 1 above)
Comment (joshlory) [reply dashboard/app/dsl/user_id_replace_dsl.rb:2]: That makes sense.  Follow up question here: https://github.com/code-dot-org/code-dot-org/pull/14486#issuecomment-295493760

**Comment by Bjvanminnen** [comment]
I wondered the same - I don't think we need to really worry about people using <user_id/> accidentally in regular External tests. Do you prefer just having this logic in External?

**Comment by joshlory** [comment]
I could see the argument either way.  But my main argument for a new level type would be if we want to customize the behavior of this new level type in the future.  If we think this is the extent of the feature I'd be fine with merging into `External`.  If we think curriculum will ask for extensions that only apply to `UserIdReplace` levels it maybe makes sense for it to be a new level type.

**Comment by Bjvanminnen** [comment]
Ok. I think I'll just add to External then.

**Comment by joshlory** [comment]
Is there any benefit to keeping the survey levels separate?  Are there future features that would only apply to `Survey` levels?

**Comment by Bjvanminnen** [comment]
I think it's hard to know that at this point in time, given that this is the only way we know we want to use it.

**Comment by Bjvanminnen** [comment]
Closing in favor of https://github.com/code-dot-org/code-dot-org/pull/14538

**Pull request closed without merging**